contracts/dao/pre-propose/cwd-pre-propose-single-overrule/src/contract.rs

```diff
@@ -103,34 +80,236 @@ pub fn execute(
         }
+        // The following messages are forwarded to the base contract
+        ExecuteMsg::ProposalCreatedHook {
+            proposal_id,
+            proposer,
+        } => Ok(ExecuteInternal::ProposalCreatedHook {
+            proposal_id,
+            proposer,
+        }),
+        ExecuteMsg::ProposalCompletedHook {
+            proposal_id,
+            new_status,
+        } => Ok(ExecuteInternal::ProposalCompletedHook {
+            proposal_id,
+            new_status,
+        }),
+        // ExecuteMsg::Withdraw and ExecuteMsg::UpdateConfig are unsupported
+        // ExecuteMsg::Withdraw is unsupported because overrule proposals should have no deposits
+        // ExecuteMsg::UpdateConfig since the config has only the info about deposits,
+        // no custom fields are added.
         _ => Err(PreProposeOverruleError::MessageUnsupported {}),
+    };
+    PrePropose::default()
+        .execute(deps, env, info, internal_msg?)
+        .map_err(PreProposeOverruleError::PreProposeBase)
+}
+
+fn get_subdao_from_timelock(
+    deps: &DepsMut,
+    timelock_contract: &String,
+) -> Result<Addr, PreProposeOverruleError> {
+    let timelock_config: TimelockTypes::Config = deps.querier.query_wasm_smart(
+        timelock_contract.to_string(),
+        &TimelockMsg::QueryMsg::Config {},
+    )?;
+    Ok(timelock_config.subdao)
+}
+
+fn get_timelock_from_subdao(
+    deps: &DepsMut,
+    subdao_core: &Addr,
+) -> Result<Addr, PreProposeOverruleError> {
+    let proposal_modules: Vec<SubdaoTypes::ProposalModule> = deps.querier.query_wasm_smart(
+        subdao_core,
+        &SubdaoQueryMsg::ProposalModules {
+            start_after: None,
+            // we assume any subdao proposal module has pre-propose module with timelock.
+            // thus, we need only single module
+            limit: Some(1),
+        },
+    )?;
+
+    if proposal_modules.is_empty() {
+        return Err(PreProposeOverruleError::SubdaoMisconfigured {});
+    }
+
+    let prop_policy: ProposalCreationPolicy = deps.querier.query_wasm_smart(
+        proposal_modules.first().unwrap().address.clone(),
+        &SubdaoProposalMsg::QueryMsg::ProposalCreationPolicy {},
+    )?;
+
+    match prop_policy {
+        ProposalCreationPolicy::Anyone {} => Err(PreProposeOverruleError::SubdaoMisconfigured {}),
+        ProposalCreationPolicy::Module { addr } => {
+            let timelock: Addr = deps.querier.query_wasm_smart(
+                addr,
+                &SubdaoPreProposeQueryMsg::QueryExtension {
+                    msg: SubdaoPreProposeQueryExt::TimelockAddress {},
+                },
+            )?;
+            Ok(timelock)
+        }
+    }
+}
+
+fn is_subdao_legit(deps: &DepsMut, subdao_core: &Addr) -> Result<bool, PreProposeOverruleError> {
+    let main_dao = get_main_dao_address(deps)?;
+
+    let mut start_after: Option<SubDao> = None;
+
+    // unfortunately, there is no way to get the total subdao number so we do infinite loop here
```

Comment:
i would consider adding another state value to the main dao with the counter of currently registered proposals. you can then query for this value to avoid the weirdness below

Reply:
do you mean add a query to check is the subdao exist?

Reply:
yes, it makes perfect sense, but my point was, if we are modifying main dao query interface, we can create a more helpful query for this especially case

Reply:
I did what @swelf19 proposed because it makes things much cleaner